Source/Core/Core/IOS/Network/Socket.cpp

```diff
@@ -712,24 +715,84 @@ void WiiSockMan::Update()
       socket_iter = WiiSockets.erase(socket_iter);
     }
   }
-  s32 ret = select(nfds, &read_fds, &write_fds, &except_fds, &t);
 
-  if (ret >= 0)
+  s32 ret = poll(ufds.data(), ufds.size(), 0);
+  if (ret < 0)
   {
-    for (auto& pair : WiiSockets)
-    {
-      WiiSocket& sock = pair.second;
-      sock.Update(FD_ISSET(sock.fd, &read_fds) != 0, FD_ISSET(sock.fd, &write_fds) != 0,
-                  FD_ISSET(sock.fd, &except_fds) != 0);
-    }
+    ERROR_LOG(IOS_NET, "SO_POLL poll failed (%d)", GetNetErrorCode(ret, "Update", true));
+    return;
   }
-  else
+
+  int n = 0;
+  for (auto& elem : WiiSockets)
   {
-    for (auto& elem : WiiSockets)
-    {
-      elem.second.Update(false, false, false);
-    }
+    elem.second.Update(!!(ufds[n].revents & read_event), !!(ufds[n].revents & write_event),
+                       !!(ufds[n].revents & error_event));
+    n++;
+  }
+
+  UpdatePollCommands();
+}
+
+void WiiSockMan::UpdatePollCommands()
+{
+  constexpr int error_event = (POLLHUP | POLLERR);
+
+  if (pending_polls.empty())
+    return;
+
+  const std::chrono::time_point<std::chrono::high_resolution_clock> now =
+      std::chrono::high_resolution_clock::now();
+  const auto elapsed_d = std::chrono::duration_cast<std::chrono::milliseconds>(now - last_time);
+  const auto elapsed = elapsed_d.count();
+  last_time = now;
+
+  for (auto& pcmd : pending_polls)
+  {
+    if (pcmd.timeout > 0)  // don't touch negative timeouts
+      pcmd.timeout = std::max(static_cast<s64>(0), pcmd.timeout - elapsed);
   }
+
+  pending_polls.erase(
+      std::remove_if(
+          pending_polls.begin(), pending_polls.end(),
+          [this](auto& pcmd) {
+            const int count = std::count_if(pcmd.wii_fds.begin(), pcmd.wii_fds.end(), [&](auto& e) {
+              if (e.revents & error_event)  // Only happens if loaded from a save state
+                return true;
+
+              auto poll_it =
+                  find_if(ufds.begin(), ufds.end(), [&](auto& u) { return e.fd == u.fd; });
+
+              // SO_POLL on IOS ignores unknown FDs instead of returning an error
+              if (poll_it == ufds.end())
+                return false;
+
+              return !!((e.events | error_event) & poll_it->revents);
+            });
+
+            if (count == 0 && pcmd.timeout)
+              return false;
+
+            const auto r = pcmd.timeout == 0 ? 0 : count;
```

Comment:
IIRC, even if poll is non-blocking, it shouldn't prevent `count` to be reported so this ternary and the `r` variable are unnecessary.